### ISSUE TITLE
Type checking a four-line function with nested unannotated dictionaries exceeds 30 seconds

### ISSUE THREAD

**Issue by temp-chatgpt-agent-oai** [astral-sh/ty#4466]
### Summary

Inferring types for two empty dictionaries used in a nested dictionary literal takes more than 30 seconds. Adding an annotation to one dictionary, or removing one dictionary layer from a conditional branch, makes checking finish in about 0.1 seconds.

These timings were reported by the original reproducer author. I have not independently rerun them.

### Minimal reproduction

```python
def f(flag: bool):
    x = {}
    y = {}
    return {"a": x, "b": {"c": y} if flag else {"d": {"e": y}}}
```

The function need not be called. Save as `repro.py` and run:

```sh
timeout 30s ty check repro.py
```

Observed: exceeds 30 seconds (timeout exit status 124).
Expected: this small function should finish checking promptly.

### Fast variants

Annotating `x` makes the same return expression finish promptly:

```python
def f(flag: bool):
    x: dict[str, object] = {}
    y = {}
    return {"a": x, "b": {"c": y} if flag else {"d": {"e": y}}}
```

Removing one layer from the false branch also avoids the slowdown:

```python
def f(flag: bool):
    x = {}
    y = {}
    return {"a": x, "b": {"c": y} if flag else {"d": y}}
```

### Versions and reported timings

Each example was checked in a separate file.

| Example | ty 0.0.78 | ty from Ruff commit `a592ffe132bc` | Pyright 1.1.408 |
| --- | --- | --- | --- |
| Original | >30 s (timeout) | >30 s (timeout) | 1.58 s |
| Annotated `x` | 0.104 s | 0.107 s | 1.56 s |
| Shallower false branch | 0.098 s | 0.101 s | 1.55 s |

The Pyright comparison used Python 3.12.9 and `{"pythonVersion": "3.12", "typeCheckingMode": "standard"}`, with no diagnostics.

### Related reports

- #3123: nested `OrderedDict` constructor slowdown (closed).
- #3663: nested literals checked against recursive `TypedDict` unions (closed).
- #4243: generic-inference blowup involving many type variables and lazy type attributes.

This reproducer has no explicit recursive types, TypedDicts, or overloaded constructor calls. It may share an underlying cause with an existing report, but I did not find an exact duplicate.
